src/supernova/mod.rs

```diff
@@ -0,0 +1,1202 @@
+
+    let digest = compute_digest::<G1, PublicParams<G1, G2>>(&[
+      running_claim1.get_publicparams(),
+      running_claim2.get_publicparams(),
+    ]);
+
+    let num_steps = rom.len();
+    let initial_program_counter = <G1 as Group>::Scalar::from(0);
+
+    // extend z0_primary/secondary with rom content
+    let mut z0_primary = vec![<G1 as Group>::Scalar::ONE];
+    z0_primary.extend(
+      rom
+        .iter()
+        .map(|opcode| <G1 as Group>::Scalar::from(*opcode as u64)),
+    );
+    let mut z0_secondary = vec![<G2 as Group>::Scalar::ONE];
+    z0_secondary.extend(
+      rom
+        .iter()
+        .map(|opcode| <G2 as Group>::Scalar::from(*opcode as u64)),
+    );
+
+    let mut recursive_snark_option: Option<RecursiveSNARK<G1, G2>> = None;
+
+    for _ in 0..num_steps {
+      let program_counter = recursive_snark_option
+        .as_ref()
+        .map(|recursive_snark| recursive_snark.program_counter)
+        .unwrap_or_else(|| initial_program_counter);
+      let augmented_circuit_index = rom[u32::from_le_bytes(
+        // convert program counter from field to usize (only took le 4 bytes)
+        program_counter.to_repr().as_ref()[0..4].try_into().unwrap(),
+      ) as usize];
+
+      let mut recursive_snark = recursive_snark_option.unwrap_or_else(|| {
+        if augmented_circuit_index == OPCODE_0 {
+          RecursiveSNARK::iter_base_step(
+            &running_claim1,
+            digest,
+            program_counter,
+            augmented_circuit_index,
+            num_augmented_circuit,
+            &z0_primary,
+            &z0_secondary,
+          )
+          .unwrap()
+        } else if augmented_circuit_index == OPCODE_1 {
+          RecursiveSNARK::iter_base_step(
+            &running_claim2,
+            digest,
+            program_counter,
+            augmented_circuit_index,
+            num_augmented_circuit,
+            &z0_primary,
+            &z0_secondary,
+          )
+          .unwrap()
+        } else {
+          unimplemented!()
+        }
+      });
+
+      if augmented_circuit_index == OPCODE_0 {
+        let res = recursive_snark.prove_step(&running_claim1, &z0_primary, &z0_secondary);
+        if let Err(e) = &res {
+          println!("res failed {:?}", e);
+        }
+        assert!(res.is_ok());
+        let res = recursive_snark.verify(&running_claim1, &z0_primary, &z0_secondary);
+        if let Err(e) = &res {
+          println!("res failed {:?}", e);
+        }
+        assert!(res.is_ok());
+      } else if augmented_circuit_index == OPCODE_1 {
+        let res = recursive_snark.prove_step(&running_claim2, &z0_primary, &z0_secondary);
+        if let Err(e) = &res {
+          println!("res failed {:?}", e);
+        }
+        assert!(res.is_ok());
+        let res = recursive_snark.verify(&running_claim2, &z0_primary, &z0_secondary);
+        if let Err(e) = &res {
+          println!("res failed {:?}", e);
```

Comment:
In a unit test, consider `.unwrap()` for errors rather than printing their contents and then asserting on ok.

Reply:
done with commit https://github.com/microsoft/Nova/pull/204/commits/61dad5b531b7316313b991d9651a32584286a6c6